firedrake/functionspaceimpl.py

```diff
@@ -0,0 +1,736 @@
+        """Return a newly allocated :class:`pyop2.MixedDat` defined on the
+        :attr:`dof_dset` of this :class:`MixedFunctionSpace`."""
+        if val is not None:
+            assert len(val) == len(self)
+        else:
+            val = [None for _ in self]
+        return op2.MixedDat(s.make_dat(v, valuetype, "%s[cmpt-%d]" % (name, i), utils._new_uid())
+                            for i, (s, v) in enumerate(zip(self._spaces, val)))
+
+    @utils.cached_property
+    def _dm(self):
+        """A PETSc DM describing the data layout for fieldsplit solvers."""
+        dm = PETSc.DMShell().create()
+        dm.setAttr('__fs__', weakref.ref(self))
+        dm.setCreateFieldDecomposition(self.create_field_decomp)
+        dm.setCreateSubDM(self.create_subdm)
+        # FIXME: The Vec could be shared between functionspaces with
+        # the same dof_dset.
+        with self.make_dat().vec_ro as v:
+            dm.setGlobalVector(v.duplicate())
+        return dm
+
+    @utils.cached_property
+    def _ises(self):
+        return self.dof_dset.field_ises
+
+    @classmethod
+    def create_subdm(cls, dm, fields, *args, **kwargs):
+        W = dm.getAttr('__fs__')()
+        if len(fields) == 1:
+            # Subspace is just a single FunctionSpace.
+            subspace = W[fields[0]]
+        else:
+            # Need to build an MFS for the subspace
+            subspace = MixedFunctionSpace([W[f] for f in fields])
+        # Sub-DM is just the DM belonging to the subspace.
+        subdm = subspace._dm
+        # Keep hold of strong reference to created subspace (given we
+        # only hold a weakref in the shell DM)
+        W._subspaces.append(subspace)
+        # Index set mapping from W into subspace.
+        iset = PETSc.IS().createGeneral(numpy.concatenate([W._ises[f].indices
+                                                           for f in fields]))
+        return iset, subdm
+
+    @classmethod
+    def create_field_decomp(cls, dm, *args, **kwargs):
+        W = dm.getAttr('__fs__')()
+        # Don't pass split number if name is None (this way the
+        # recursively created splits have the names you want)
+        names = [s.name for s in W]
+        dms = [V._dm for V in W]
+        return names, W._ises, dms
+
+
+class ProxyFunctionSpace(FunctionSpace):
+    """A :class:`FunctionSpace` that one can attach extra properties to.
+
+    :arg mesh: The mesh to use.
+    :arg element: The UFL element.
+    :arg name: The name of the function space.
+
+    .. warning::
+
+       Users should not build a :class:`ProxyFunctionSpace` directly,
+       it is mostly used as an internal implementation detail.
+    """
+    def __new__(cls, mesh, element, name=None):
+        topology = mesh.topology
+        self = super(ProxyFunctionSpace, cls).__new__(cls, topology, element, name=name)
+        if mesh is not topology:
+            return WithGeometry(self, mesh)
+        else:
+            return self
```

Comment:
Could remove need for new by making a `ProxyFunctionSpaceBase` and making `ProxyFunctionSpace` a function not a class.  But it's not too complicated, so I don't think there's a need.  Note that argument sanitisation doesn't take place here.  The use case is, there is an existing `FunctionSpace` and now we need to add some properties.